### PR TITLE
Fix Picture in Picture incorrectly stopped when no video view is initially available

### PR DESCRIPTION
**Pull request by defagos**
# Description

The problem reported in the associated issue is related to the fact that no player layer might be initially available. This is something that we cannot control reliably, as implementations might choose to display no video view initially.

I considered two fixes, both as last commits on the associated branch:

- Perform reference counting without checking if a player layer is available.
- Perform reference counting as before when a player layer is available, but perform separate pending reference counting when no player layer is available.

I think the first option should be enough, except if two video views are displayed on a parent view which is enabled for in-app PiP (in which case reference counting might be interleaved).

Once the above fix was made I discovered another issue with the sample code provided in the linked issue. PiP works but is paused when changing the content. It namely appears that updating the player item must be made after PiP will stop method is called. To achieve this result I had to defer setup of the new item, which can only be made by introducing a setup block to the in-app PiP modifier. This block makes it possible to perform the setup after PiP is stopped, with a very small delay required, though.

# Remark

I'll put this PR in review (without auto-merge) but I think we should:

- Discuss the fix first. Implementing in-app PiP for the multiple view demo might help the discussion.
- Test the behavior on an Apple TV device (which I don't have).

# Changes made

- Fix PiP bug. 
- Update the in-app PiP modifier to take a setup closure.
- Remove player layout (makes things more complicated, especially since the system player layout can be tested under the Showcase tab).
- Remove `WillAppearView`, not required anymore.
- Update all demo examples.

# Checklist

- [x] APIs have been properly documented (if relevant).
- [x] The documentation has been updated (if relevant).
- [x] New unit tests have been written (if relevant).
- [x] The demo has been updated (if relevant).
- [x] The playground has been updated (if relevant).

### CODE REVIEW

File: Sources/Player/UserInterface/View.swift [hunk 1]
```diff
@@ -34,12 +34,9 @@ public extension View {
     /// required by Picture in Picture initiated from the view are not needed anymore.
     ///
     /// - Parameter cleanup: A closure to clean resources associated with the view.
-    func enabledForInAppPictureInPictureWithCleanup(perform cleanup: @escaping () -> Void) -> some View {
+    func enabledForInAppPictureInPictureWithSetup(perform setup: @escaping () -> Void, cleanup: @escaping () -> Void) -> some View {
```
Comment (MGaetan89) [inline Sources/Player/UserInterface/View.swift:37]: Don't you need to update the comment of this method?

**Comment by waliid** [comment]
Thank you for this PR. However, I wonder if we are not going too far.

1. Is it an edge case?
2. Should we try to fix this issue?
3. Shouldn't we always have a `VideoView`, regardless of the media type?

I don't see a situation where we wouldn't want a `VideoView`.
If I want to play content, regardless of its state, I need a `VideoView`.

It seems to me that we are addressing an edge case, but I may be wrong.

**Comment by defagos** [comment]
> 1. Is it an edge case?

If your implementation always has a `VideoView` then, yes, this is an edge case (this is why it was not discovered until now).

> 2. Should we try to fix this issue?
> 3. Shouldn't we always have a `VideoView`, regardless of the media type?

We have no control over how the player will be integrated, especially not if a `VideoView` is guaranteed to be available when the PiP enabled view appears for the first time.

The problem is that clients integrating the library will see PiP fail for no reason. Check the [sample code](https://github.com/SRGSSR/pillarbox-apple/issues/696#issue-2042187883). How is it possible that such a small change can make PiP fail? Clients will never be able to understand and debug the issue.

One possibility was mentioned in the associated issue: Make the modifiers be applied on `VideoView` or `SystemVideoView` itself. This way both can never be separated. But sadly this approach does not work either since it couples the lifetime of PiP too tightly to the video view, leading to other issues.

If we think this is going too far we can:

- Do nothing, but this issue will arise.
- Force a `VideoView` to be always available when PiP reference counting takes place. This could be achieved with an assertion, though in this case there is no guarantee that the issue will be discovered early in client code (not all layout possibilities might be tested well enough in client code).

There are also two implementations possible, as mentioned in the PR description. While the second one is maybe too much, the first one is actually shorter than our initial implementation, since it removes all layer identity tests.

File: Sources/Player/UserInterface/View.swift (hunk 1 above)
Comment (defagos) [reply Sources/Player/UserInterface/View.swift:39]: Fixed, thanks.

**Comment by defagos** [comment]
There is no hurry in merging this PR. I discovered the issue when implementing 360° support (there is a video view, but it cannot be enabled for PiP). I'll have a look at other options.

**Comment by defagos** [comment]
I moved this PR to draft.

I am strongly convinced we have to address this issue. We cannot rely on users correctly inserting a video view in PiP-enabled views, this can break too easily when refactoring SwiftUI code, something quite common (and I was already bitten several times).

The solution of this PR is not good enough, though. It involves API changes (setup phase) which are likely not what we want. I will try to find an elegant solution instead.

**Comment by defagos** [comment]
Found a way to make our current PiP implementation simpler and to solve this issue. This work has been done alongside 360° integration since a complete implementation requires both issues to be solved. I'll open the dedicated PR later today.

This PR can be forgotten. Closed.

**Pull request closed without merging**